operator/pkg/lbipam/lbipam.go

```diff
@@ -0,0 +1,1673 @@
+}
+
+type LBIPAMInitDone chan struct{}
+
+type LBIPAMParams struct {
+	cell.In
+
+	Logger       logrus.FieldLogger
+	LC           hive.Lifecycle
+	ClientSet    k8sClient.Clientset
+	PoolResource resource.Resource[*cilium_api_v2alpha1.CiliumLoadBalancerIPPool]
+	SvcResource  resource.Resource[*core_v1.Service]
+	Config       LBIPAMConfig
+	InitDone     LBIPAMInitDone `optional:"true"`
+}
+
+func registerLBIPAM(params LBIPAMParams) *LBIPAM {
+	if !params.Config.EnableLBIPAM {
+		return nil
+	}
+
+	var lbClasses []string
+	if operatorOption.Config.BGPAnnounceLBIP {
+		lbClasses = append(lbClasses, "io.cilium/bgp-control-plane")
+	}
+
+	lbIPAM := &LBIPAM{
+		Logger:       params.Logger,
+		PoolResource: params.PoolResource,
+		SvcResource:  params.SvcResource,
+		InitDone:     params.InitDone,
+		PoolStore:    NewPoolStore(),
+		RangesStore:  NewRangesStore(),
+		ServiceStore: NewServiceStore(),
+		LBClasses:    lbClasses,
+		IPv4Enabled:  option.Config.IPv4Enabled(),
+		IPv6Enabled:  option.Config.IPv6Enabled(),
+	}
+
+	// Make a new Ctx which we will use to manage the lifetime of LBIPAM.Run.
+	ipamCtx, cancel := context.WithCancel(context.Background())
+	ipamShutdown := make(chan struct{})
+
+	params.LC.Append(hive.Hook{
+		OnStart: func(ctx context.Context) error {
+			params.Logger.Info("Starting LB IPAM")
+
+			// Add the clients at start time since the client set doesn't have the clients available during
+			// initialization just yet.
+			lbIPAM.PoolClient = params.ClientSet.CiliumV2alpha1().CiliumLoadBalancerIPPools()
+			lbIPAM.SvcClient = params.ClientSet.CoreV1()
+
+			// Start go routine, we can't block in OnStart
+			go func() {
+				lbIPAM.Run(ipamCtx)
+				close(ipamShutdown)
+			}()
+
+			return nil
+		},
+		OnStop: func(ctx context.Context) error {
+			// Close the ipamCtx
+			cancel()
+
+			// Wait for LB IPAM shutdown to be complete or the Stop ctx to expire.
+			select {
+			case <-ipamShutdown:
+			case <-ctx.Done():
+			}
+
+			return nil
+		},
+	})
+
+	return lbIPAM
+}
+
+// LBIPAM is the loadbalancer IP address manager, watcher/controller which allocates and assigns IP addresses
+// to LoadBalancer services from the configured set of LoadBalancerIPPools in the cluster.
+type LBIPAM struct {
+	Logger logrus.FieldLogger
+
+	PoolClient cilium_client_v2alpha1.CiliumLoadBalancerIPPoolInterface
+	SvcClient  client_typed_v1.ServicesGetter
+
+	PoolResource resource.Resource[*cilium_api_v2alpha1.CiliumLoadBalancerIPPool]
+	PoolSynced   bool
+
+	SvcResource resource.Resource[*core_v1.Service]
+	SvcSynced   bool
+
+	PoolStore    PoolStore
+	RangesStore  RangesStore
+	ServiceStore ServiceStore
+
+	LBClasses   []string
+	IPv4Enabled bool
+	IPv6Enabled bool
+
+	// Only used during testing.
+	InitDone LBIPAMInitDone
+}
+
+func (ipam *LBIPAM) Run(ctx context.Context) {
+	//
+	errChan := make(chan error, 2)
+
+	poolChan := stream.ToChannel[resource.Event[*cilium_api_v2alpha1.CiliumLoadBalancerIPPool]](ctx, errChan, ipam.PoolResource)
```

Comment:
btw, just so you're aware, `ToChannel` is currently unbuffered because I wasn't sure what the default should be. We likely want it unbuffered so that we don't have items in the channel's buffer but rather in the resource's queue (which get deduplicated).